### DISCUSSION TITLE
How to reset globals after using stubGlobal?

### DISCUSSION THREAD

**Discussion by bwiggs** [vitest-dev/vitest#1741] [Q&A] [answered]
What's the best way to reset the global window object after I've set a bunch of stubs with `vi.stubGlobal`?

I'm working on some helper utilities that set a bunch of global state on the window object. After each test I'd like to reset the global window object to a clean state. Currently I'm tracking all the properties that I know if that have been set with  `stubGlobal`, but that doesn't cover the case that someone adds a their own globals. My understanding is that these would then be leaked to other tests?

**Comment by sheremet-va** [comment, marked as answer]
It's impossible to trackdown. This is why we provide isolation with workers, and jest provides isolation by created a new global object for each test.

**Reply by bwiggs** [reply]
@sheremet-va thank you!

It looks like I get a new global object for each test _file_, as opposed to each test?

> Vitest also isolates each file's environment so env mutations in one file don't affect others. Isolation can be disabled by passing --no-isolate to the CLI (trading correctness for run performance).

So I shouldn't expect globals to leak into tests from one file to another, but in a single file I could get leaks between tests

**Reply by sheremet-va** [reply]
Yes, this is correct.

**Reply by bwiggs** [reply]
Got it, thank you for the clarification.

**Comment by joranminjon88** [comment]
For anyone reaching this thread: unstubbing globals is possible with: `vi.unstubAllGlobals()`. I am running this after each test and nothing stubbed with `vi.stubGlobal` leaks to the next test. This function has been in vitest since v1.

**Reply by joranminjon88** [reply]
This of course prevents you from stubbing a global function, well, globally in the file. Setting a global function on root level of your test suite works only for the first test - when using afterEach. After that the global stub is unstubbed and needs to be set explicitly in each test. You may or may not want that.
